### ISSUE TITLE
Are use of `dlopen` necessary

### ISSUE THREAD

**Issue by jeswr** [SWI-Prolog/swipl-devel#1158]
Per https://github.com/emscripten-core/emscripten/issues/19141#issuecomment-1498443649 - `dlopen` usages appear to be causing problems with the webassembly builds of this library - I'm wondering if they can be removed.

**Comment by JanWielemaker** [comment]
Surely, if there are calls to dlopen, they can be `#ifdef'ed` for the STATIC_EXTENSIONS setup.

**Comment by jeswr** [comment]
Perhaps its an extra conditional required here (https://github.com/SWI-Prolog/swipl-devel/blob/13d22dd197a89df4c2f33dc21da612c2d6dd85a6/src/pl-load.c#L110)?

The full list of `dlopen` references I can find is at https://github.com/search?q=repo%3ASWI-Prolog%2Fswipl-devel%20dlopen&type=code.

I'd offer to help but unfortunately I don't have enough experience with the core codebase to do it in a reasonable amount of time - and I am generally time poor at the moment.

**Comment by JanWielemaker** [comment]
Yes.   Part of the code is already disabled under `O_STATIC_EXTENSIONS` and replaced  by other code (to call the init functions of the statically added components).    Propably we should simply disable more.    We do not want all the predicates for loading and unloading shared objects and their helpers.

**Comment by dmchurch** [comment]
I think there was some sort of handling of dlopen on my emscripten branch?
you might check, at least.

**Comment by JanWielemaker** [comment]
> I think there was some sort of handling of dlopen on my emscripten branch?

Considering the warnings against dynamic loading in the Emscripten docs I tried a different route, allowing extensions to be linked statically.  That is the `O_STATIC_EXTENSIONS`.    It compiles the extension to a static lib, generates a C file with a data structure that links the library name to its init function and links the whole thing statically.    Next, "linking" such a library using use_foreign_library/1 simply looks up and calls the init function.

Possibly we want that different at some point.  For now, this is fairly simple and the main disadvantage is that it is not so easy to get user extensions into this framework.     Ideally we'd move such extensions to use CMake with some cmake include file provided by Prolog that allows this to work.

**Comment by JanWielemaker** [comment]
The WASM build uses O_STATIC_EXTENSIONS which disables dlopen and produces a working binary. Closing.
